homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,506 @@
+        last_recurrence_start_datetime = self.rrule.before(now, inc=True)
+        last_recurrence_end_datetime = last_recurrence_start_datetime + self._duration
+        return last_recurrence_start_datetime <= now <= last_recurrence_end_datetime
+
+    def as_dict(self) -> dict:
+        """Return the schedule as a dict."""
+        the_dict = {
+            CONF_ENTITY_ID: self.entity_id,
+            CONF_START_DATETIME: self.start_datetime.isoformat(),
+        }
+
+        if self.end_datetime:
+            the_dict[CONF_END_DATETIME] = self.end_datetime.isoformat()
+        else:
+            the_dict[CONF_END_DATETIME] = None
+
+        if self.rrule:
+            the_dict[CONF_RRULE] = str(self.rrule)
+        else:
+            the_dict[CONF_RRULE] = None
+
+        return the_dict
+
+    async def async_schedule_next_instance(self) -> None:
+        """Schedule the next instance of a schedule."""
+        now = datetime.now()
+
+        # Calculate the start and end datetimes for the next instance:
+        if not self._initial_run_complete:
+            self._initial_run_complete = True
+            start_datetime = self.start_datetime
+            end_datetime = self.end_datetime
+        elif self.rrule:
+            start_datetime = self.rrule.after(now, inc=True)
+            if self.end_datetime:
+                end_datetime = start_datetime + self._duration
+            else:
+                end_datetime = None
+
+        if not start_datetime:
+            _LOGGER.debug("No more instances of schedule: %s", self)
+            return
+
+        self.active_instance: ActiveScheduleInstance = ActiveScheduleInstance(
+            self._hass,
+            self.async_schedule_next_instance,
+            self.entity_id,
+            start_datetime,
+            end_datetime=end_datetime,
+        )
+        self.active_instance.async_init()
+
+        _LOGGER.info(
+            "Scheduled instance of schedule %s (start: %s / end: %s)",
+            self,
+            start_datetime,
+            end_datetime,
+        )
+
+
+@bind_hass
+class Scheduler:
+    """A class to manage scene schedules."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize."""
+        self._hass: HomeAssistant = hass
+        self._latest_instances: List[ActiveScheduleInstance] = []
+        self._store = Store(hass, STORAGE_VERSION, DOMAIN)
+        self.schedules: Dict[str, Schedule] = {}
+
+    async def async_clear_expired_schedules(self) -> None:
+        """Delete expired schedules."""
+        now = datetime.now()
+        self.schedules = {
+            schedule_id: schedule
+            for schedule_id, schedule in self.schedules.items()
+            if (not schedule.end_datetime and schedule.start_datetime < now)
+            or (schedule.end_datetime and schedule.end_datetime < now)
+        }
+
+        await self.async_save()
+
+    async def async_create(
+        self,
+        entity_id: str,
```

Comment:
Just have it take a `Schedule` as parameter. Otherwise you'll forever have to keep this function in sync with `Schedule`, that is too tightly coupled